controllers/update_pods.go

```diff
@@ -182,6 +199,15 @@ func getPodsToUpdate(logger logr.Logger, reconciler *FoundationDBClusterReconcil
 		updates[zone] = append(updates[zone], pod)
 	}
 
+	var numPodsToUpdate int
+	for _, zone := range updates {
+		numPodsToUpdate += len(zone)
+	}
+
+	if maxUnavailablePods > 0 && (unavailablePods+numPodsToUpdate) >= maxUnavailablePods {
```

Comment:
I believe this calculation is still wrong. Assume a user sets `maxUnavailablePods` to `1` to make sure only 1 Pod is unavailable at a time (at least from planned activities) and now the user makes a change to the cluster that requires the operator to update all Pods in the cluster. For simplicity assume the cluster has 3 Pods, in this case the operator will never be able to make progress as `(unavailablePods+numPodsToUpdate) >= maxUnavailablePods` will be `3 >= 1` which will be blocking forever.

I believe a better approach is to return the updates map and the number of Pods that can be updated, in the case above that would be 1 if all Pods are available. If the updates map has at lease one entry but the Pods that can be updated is 0 (or a negative number) the operator should do a (delayed) requeue. If the number of Pods that can be updated is greater than 0 the operator should prefer to update zones that have less or equal the number of Pods that can be updated (to make sure we update the zone at once) if not the operator should update as many Pods as possible from the zone.

That way we make sure the operator is able to make progress in the different cases.

Reply:
That makes total sense. I like the idea. I hope I got it right this time :D